go/oasis-node/cmd/common/metrics/metrics.go

```diff
@@ -179,56 +312,533 @@ func (s *pushService) worker() {
 	switch strings.ToLower(mode) {
-	case metricsModeNone:
+	case MetricsModeNone:
 		return newStubService()
-	case metricsModePull:
+	case MetricsModePull:
 		return newPullService(ctx)
-	case metricsModePush:
+	case MetricsModePush:
 		return newPushService()
 	default:
 		return nil, fmt.Errorf("metrics: unsupported mode: '%v'", mode)
 	}
 }
 
+type diskService struct {
+	service.BaseBackgroundService
+	sync.Mutex
+
+	dataDir string
+	// TODO: Should we monitor I/O of children PIDs as well?
+	pid      int
+	interval time.Duration
+
+	diskUsageGauge          prometheus.Gauge
+	diskIOReadBytesGauge    prometheus.Gauge
+	diskIOWrittenBytesGauge prometheus.Gauge
+}
+
+func (d *diskService) Start() error {
+	go d.worker()
+	return nil
+}
+
+// updateDiskUsage walks through dataDir and updates diskUsageGauge attribute.
+func (d *diskService) updateDiskUsage() error {
+	d.Lock()
+	defer d.Unlock()
+
+	// Compute disk usage of datadir.
+	var duBytes int64
+	err := filepath.Walk(d.dataDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("disk usage metric: failed to access file %s: %v", path, err)
+		}
+		duBytes += info.Size()
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("disk usage metric: failed to walk directory %s: %v", d.dataDir, err)
+	}
+	d.diskUsageGauge.Set(float64(duBytes))
+
+	return nil
+}
+
+// updateDiskUsage reads process info and updates diskIO{Read|Written}BytesGauge attributes.
+func (d *diskService) updateIO() error {
+	d.Lock()
+	defer d.Unlock()
+
+	// Obtain process I/O info.
+	proc, err := procfs.NewProc(d.pid)
+	if err != nil {
+		return fmt.Errorf("disk I/O metric: failed to obtain proc object for PID %d: %v", d.pid, err)
+	}
+	procIO, err := proc.IO()
+	if err != nil {
+		return fmt.Errorf("disk I/O metric: failed to obtain procIO object %d: %v", d.pid, err)
+	}
+
+	d.diskIOWrittenBytesGauge.Set(float64(procIO.ReadBytes))
+	d.diskIOReadBytesGauge.Set(float64(procIO.WriteBytes))
+
+	return nil
+}
+
+func (d *diskService) worker() {
+	t := time.NewTicker(d.interval)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-d.Quit():
+			break
+		case <-t.C:
+		}
+
+		if err := d.updateDiskUsage(); err != nil {
+			d.Logger.Warn(err.Error())
+		}
+
+		if err := d.updateIO(); err != nil {
+			d.Logger.Warn(err.Error())
+		}
+	}
+}
+
+// NewDiskService constructs a new disk usage and I/O service.
+//
+// This service will compute the size of datadir folder and read I/O info of the process every --metric.push.interval
+// seconds.
+func NewDiskService() (service.BackgroundService, error) {
+	ds := &diskService{
+		BaseBackgroundService: *service.NewBaseBackgroundService("disk"),
+		dataDir:               viper.GetString(common.CfgDataDir),
+		pid:                   os.Getpid(),
+		interval:              viper.GetDuration(CfgMetricsPushInterval),
+
+		diskUsageGauge: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Name: "oasis_worker_disk_usage_bytes",
```

Comment:
Move metric names into constants so the names can be referenced from the `ba` tool